### DISCUSSION TITLE
Why do "alerts" and "#select" have different result

### DISCUSSION THREAD

**Discussion by kal-purush** [github/codeql#10525] [Q&A]
I am running this query on vscode

```
import javascript
from DataFlow::InvokeNode call, Function caller, Function callee
where caller = call.getEnclosingExpr().getEnclosingFunction()
  and callee = call.getACallee()
select caller, caller.getFile().toString()+":"+callee.getFile().toString() 
```

I got **247** results if I choose "#select" in the query result panel but only **163** if I choose "alerts"? Why do these results vary?

**Comment by edoardopirovano** [comment]
Thanks for getting in touch with this question! The `#select` option will return the results of evaluating the query on the CodeQL database without trying to correlate them back to the source code you are analysing. The `alerts` view on the other hand will interpret the results of the query and attempt to map them back to the source code being analyzed. In particular, this means that some results in dependencies that we still included the code for in the database in order to make the analysis more complete but weren't actually part of the sources being analysed will be dropped when switching from `#select` to `alerts`. This behaviour is deliberate as you typically do not want to see alerts for this code that isn't part of your project.

If there's results being dropped that you would expect to be included, I'm happy to look into that for you but I'll need a few more details about how you obtained the database you're querying. Did you create it yourself with the CLI? If so with what flags? In particular, the `--source-root` option passed to `codeql database create` is the important one here, as that defines where the sources that we actually care about and want to keep alerts for are located.
